### ISSUE TITLE
Calligraphy for closed paths

### ISSUE THREAD

**Issue by kpym** [loopspace/spath3#3]
Is it normal that I can't use `calligraphy` on closed paths ?

For example `\calligraphy (0,0) arc(0:360:1);` is ok but `\calligraphy (0,0) arc(0:360:1) -- cycle;` or `\calligraphy (0,0) circle(1);` throws `\l__spath_tmpc_tl=\pgfsyssoftpath@closepathtoken . <recently read> }`.

**Comment by loopspace** [comment]
It's certainly a bug, but one that I was leaving until someone actually noticed it.

It relates to the fact that in order to build the thickened calligraphy path then the library needs to reverse the existing path.  This is mildly more complicated for a path with a `closeto` in it than a path without, so I hadn't implemented that.

In my defence, the calligraphy mechanism works best when the path doesn't double back on itself (ie switch from going left-to-right to going right-to-left and similarly for vertical) so this isn't a frequent use-case.

Nevertheless, it's a bug so I'll take a look.

**Comment by kpym** [comment]
I agree that this is not a big problem. I see two "easy" possible corrections.

- Simply output a more readable message when the path is closed. Something like "Closed paths are not allowed in calligraphy.", or
- Replace all `closeto` by ` moveto`.

In any case I do not need this, I just wanted to inform you.

**Comment by loopspace** [comment]
I think this is fixed in ad3e64229dc80a6583e5cbe6f11fdd612408b9e7

The `circle` path is a curious one.  A circle *by itself* contains two successive `moveto`s, since when `\draw (0,0) circle[radius=1];` is parsed, the coordinate `(0,0)` triggers a `moveto` to `(0,0)`, but then the circle is drawn from a point on the circumference so there's an additional `moveto` to a second point.

In the `calligraphy` library, a path component consisting only of a `moveto`s is still considered part of the path (this is particularly important in the `pen` path), so with a circle it looks as if the pen has been momentarily placed at the centre of the circle.  Maybe for the drawn path, this is not the expected behaviour and components consisting of only a `moveto` should not be used.

I'll need to ponder that one a bit.

**Comment by kpym** [comment]
I don't know the internals but for me `moveto` is "start a new path". In the `pen` cas `moveto` looks likes "add a new pen to the existing one" (and then we loop over all this pens to produce the calligraphic paths), in the `path`case it looks for me like "finish the current path and add one new calligraphic path for any component of the pen, then start a new path (that will be calligraphied)". 

In other words I was thinking that you loop over all components (delimited by `moveto`) of a path, and then for each component you loop over all parts (delimited by `moveto`) of the pen to add a new calligraphic path.

**Comment by loopspace** [comment]
That's right.  The caveat is that with the pen, a lone coordinate specifies a copperplate-type pen.  So a pen path of the form `(0,0) -- (.25,.25) (.3,.3)` will model a *split nib* style nib with a thick lower part and a thin upper part.  So I'm keeping that behaviour.

However, I agree that it doesn't make sense with the path itself so I've modified the behaviour to ignore segments that consist only of a `moveto`.

Fixed in: b94b856e0b4aa009237fad5f1a9823a95bb400c5

**Comment by kpym** [comment]
Circle looks good when the pen is a "real" path, but is a little bit strange in the case `\pen (0,0);\calligraphy circle(1);`.

![\pen (0,0);\calligraphy circle(1);](https://i.imgur.com/aUlT7YP.png)

**Comment by loopspace** [comment]
That's the `closepath` in the circle interfering with the copperplate pen's tapering mode.  A closepath doesn't make sense here, so it should be removed (or replaced by a lineto if of non-zero length).

**Comment by kpym** [comment]
Just for info `\pen (0,0);\calligraphy arc(0:360:1) -- cycle;`  is different from `\pen (0,0);\calligraphy (0:1) arc(0:360:1) -- cycle;` which is equivalent to `\calligraphy circle(1);`. So it looks like the problems is in the `moveto` .

**Comment by loopspace** [comment]
I believe that this is now fixed (but I've lost track of which commit it was fixed in).  See the file `calligraphy_closed.tex` for examples demonstrating it working.
